File: APIs/javascript/graphql_api.js
const { ApolloServer, gql } = require('apollo-server');

const usuarios_db = [];
const musicas_db = [];
const playlists_db = [];
const playlist_musica_db = [];

const typeDefs = gql`
  type Usuario { id: ID!, nome: String!, idade: Int!, playlists: [Playlist!]! }
  type Musica { id: ID!, nome: String!, artista: String! }
  type Playlist { id: ID!, nome: String!, usuario: Usuario!, musicas: [Musica!]! }

  type Query {
    listarUsuarios: [Usuario!]!
    obterUsuario(id: ID!): Usuario
    listarMusicas: [Musica!]!
    obterMusica(id: ID!): Musica
    listarPlaylists: [Playlist!]!
    obterPlaylist(id: ID!): Playlist
    listarPlaylistsPorUsuario(usuarioId: ID!): [Playlist!]!
    listarMusicasPorPlaylist(playlistId: ID!): [Musica!]!
    listarPlaylistsPorMusica(musicaId: ID!): [Playlist!]!
  }

  type Mutation {
    criarUsuario(id: ID!, nome: String!, idade: Int!): Usuario!
    atualizarUsuario(id: ID!, nome: String, idade: Int): Usuario!
    deletarUsuario(id: ID!): Boolean!

    criarMusica(id: ID!, nome: String!, artista: String!): Musica!
    atualizarMusica(id: ID!, nome: String, artista: String): Musica!
    deletarMusica(id: ID!): Boolean!

    criarPlaylist(id: ID!, nome: String!, usuarioId: ID!): Playlist!
    atualizarPlaylist(id: ID!, nome: String): Playlist!
    deletarPlaylist(id: ID!): Boolean!

    adicionarMusicaPlaylist(playlistId: ID!, musicaId: ID!): Playlist!
    removerMusicaPlaylist(playlistId: ID!, musicaId: ID!): Playlist!
  }
`;

const resolvers = {
  Query: {
    listarUsuarios: () => usuarios_db,
    obterUsuario: (_, { id }) => usuarios_db.find(u => u.id === id) || null,
    listarMusicas: () => musicas_db,
    obterMusica: (_, { id }) => musicas_db.find(m => m.id === id) || null,
    listarPlaylists: () => playlists_db,
    obterPlaylist: (_, { id }) => playlists_db.find(p => p.id === id) || null,
    listarPlaylistsPorUsuario: (_, { usuarioId }) => playlists_db.filter(p => p.usuario_id === usuarioId),
    listarMusicasPorPlaylist: (_, { playlistId }) => {
      const m_ids = playlist_musica_db.filter(pm => pm[0] === playlistId).map(pm => pm[1]);
      return musicas_db.filter(m => m_ids.includes(m.id));
    },
    listarPlaylistsPorMusica: (_, { musicaId }) => {
      const p_ids = playlist_musica_db.filter(pm => pm[1] === musicaId).map(pm => pm[0]);
      return playlists_db.filter(p => p_ids.includes(p.id));
    }
  },
  Mutation: {
    criarUsuario: (_, { id, nome, idade }) => {
      if (usuarios_db.find(u => u.id === id)) throw new Error("Usuário já existe");
      const u = { id, nome, idade };
      usuarios_db.push(u);
      return u;
    },
    atualizarUsuario: (_, { id, nome, idade }) => {
      const u = usuarios_db.find(u => u.id === id);
      if (!u) throw new Error("Usuário não encontrado");
      if (nome) u.nome = nome;
      if (idade) u.idade = idade;
      return u;
    },
    deletarUsuario: (_, { id }) => {
      const idx = usuarios_db.findIndex(u => u.id === id);
      if (idx === -1) throw new Error("Usuário não encontrado");
      usuarios_db.splice(idx, 1);
      playlists_db.splice(0, playlists_db.length, ...playlists_db.filter(p => p.usuario_id !== id));
      return true;
    },

    criarMusica: (_, { id, nome, artista }) => {
      if (musicas_db.find(m => m.id === id)) throw new Error("Música já existe");
      const m = { id, nome, artista };
      musicas_db.push(m);
      return m;
    },
    atualizarMusica: (_, { id, nome, artista }) => {
      const m = musicas_db.find(m => m.id === id);
      if (!m) throw new Error("Música não encontrada");
      if (nome) m.nome = nome;
      if (artista) m.artista = artista;
      return m;
    },
    deletarMusica: (_, { id }) => {
      const idx = musicas_db.findIndex(m => m.id === id);
      if (idx === -1) throw new Error("Música não encontrada");
      musicas_db.splice(idx, 1);
      playlist_musica_db.splice(0, playlist_musica_db.length, ...playlist_musica_db.filter(pm => pm[1] !== id));
      return true;
    },

    criarPlaylist: (_, { id, nome, usuarioId }) => {
      if (!usuarios_db.find(u => u.id === usuarioId)) throw new Error("Usuário não existe");
      if (playlists_db.find(p => p.id === id)) throw new Error("Playlist já existe");
      const p = { id, nome, usuario_id: usuarioId };
      playlists_db.push(p);
      return p;
    },
    atualizarPlaylist: (_, { id, nome }) => {
      const p = playlists_db.find(p => p.id === id);
      if (!p) throw new Error("Playlist não encontrada");
      if (nome) p.nome = nome;
      return p;
    },
    deletarPlaylist: (_, { id }) => {
      const idx = playlists_db.findIndex(p => p.id === id);
      if (idx === -1) throw new Error("Playlist não encontrada");
      playlists_db.splice(idx, 1);
      playlist_musica_db.splice(0, playlist_musica_db.length, ...playlist_musica_db.filter(pm => pm[0] !== id));
      return true;
    },

    adicionarMusicaPlaylist: (_, { playlistId, musicaId }) => {
      if (!playlists_db.find(p => p.id === playlistId)) throw new Error("Playlist não encontrada");
      if (!musicas_db.find(m => m.id === musicaId)) throw new Error("Música não encontrada");
      if (playlist_musica_db.some(pm => pm[0] === playlistId && pm[1] === musicaId)) {
        throw new Error("Música já existe na playlist");
      }
      playlist_musica_db.push([playlistId, musicaId]);
      return playlists_db.find(p => p.id === playlistId);
    },
    removerMusicaPlaylist: (_, { playlistId, musicaId }) => {
      if (!playlists_db.find(p => p.id === playlistId)) throw new Error("Playlist não encontrada");
      if (!musicas_db.find(m => m.id === musicaId)) throw new Error("Música não encontrada");
      const idx = playlist_musica_db.findIndex(pm => pm[0] === playlistId && pm[1] === musicaId);
      if (idx === -1) throw new Error("Música não está na playlist");
      playlist_musica_db.splice(idx, 1);
      return playlists_db.find(p => p.id === playlistId);
    }
  },
  Usuario: { playlists: (parent) => playlists_db.filter(p => p.usuario_id === parent.id) },
  Playlist: {
    usuario: (parent) => usuarios_db.find(u => u.id === parent.usuario_id),
    musicas: (parent) => {
      const m_ids = playlist_musica_db.filter(pm => pm[0] === parent.id).map(pm => pm[1]);
      return musicas_db.filter(m => m_ids.includes(m.id));
    }
  }
};

const server = new ApolloServer({ typeDefs, resolvers });

server.listen({ port: 9002 }).then(({ url }) => {
  console.log(`GraphQL API (Node.js) ready at ${url}`);
});

File: APIs/python/graphql_api.py
import strawberry
from fastapi import FastAPI
from typing import List, Optional

usuarios_db = []
musicas_db = []
playlists_db = []
playlist_musica_db = []

@strawberry.type
class Musica:
    id: strawberry.ID
    nome: str
    artista: str

@strawberry.type
class Playlist:
    id: strawberry.ID
    nome: str
    usuario_id: str

    @strawberry.field
    def musicas(self) -> List[Musica]:
        m_ids = [pm[1] for pm in playlist_musica_db if pm[0] == self.id]
        return [m for m in musicas_db if m.id in m_ids]

@strawberry.type
class Usuario:
    id: strawberry.ID
    nome: str
    idade: int

    @strawberry.field
    def playlists(self) -> List[Playlist]:
        return [p for p in playlists_db if p.usuario_id == self.id]

@strawberry.type
class Query:
    @strawberry.field
    def listar_usuarios(self) -> List[Usuario]:
        return usuarios_db

    @strawberry.field
    def obter_usuario(self, id: strawberry.ID) -> Optional[Usuario]:
        for u in usuarios_db:
            if u.id == id:
                return u
        return None

    @strawberry.field
    def listar_musicas(self) -> List[Musica]:
        return musicas_db

    @strawberry.field
    def obter_musica(self, id: strawberry.ID) -> Optional[Musica]:
        for m in musicas_db:
            if m.id == id:
                return m
        return None

    @strawberry.field
    def listar_playlists(self) -> List[Playlist]:
        return playlists_db

    @strawberry.field
    def obter_playlist(self, id: strawberry.ID) -> Optional[Playlist]:
        for p in playlists_db:
            if p.id == id:
                return p
        return None

    @strawberry.field
    def listar_playlists_por_usuario(self, usuario_id: strawberry.ID) -> List[Playlist]:
        return [p for p in playlists_db if p.usuario_id == usuario_id]

    @strawberry.field
    def listar_musicas_por_playlist(self, playlist_id: strawberry.ID) -> List[Musica]:
        m_ids = [pm[1] for pm in playlist_musica_db if pm[0] == playlist_id]
        return [m for m in musicas_db if m.id in m_ids]

    @strawberry.field
    def listar_playlists_por_musica(self, musica_id: strawberry.ID) -> List[Playlist]:
        p_ids = [pm[0] for pm in playlist_musica_db if pm[1] == musica_id]
        return [p for p in playlists_db if p.id in p_ids]

@strawberry.type
class Mutation:
    @strawberry.mutation
    def criar_usuario(self, id: strawberry.ID, nome: str, idade: int) -> Usuario:
        if any(u.id == id for u in usuarios_db):
            raise Exception("Usuário já existe")
        u = Usuario(id=id, nome=nome, idade=idade)
        usuarios_db.append(u)
        return u

    @strawberry.mutation
    def atualizar_usuario(self, id: strawberry.ID, nome: Optional[str] = None, idade: Optional[int] = None) -> Usuario:
        for u in usuarios_db:
            if u.id == id:
                if nome is not None:
                    u.nome = nome
                if idade is not None:
                    u.idade = idade
                return u
        raise Exception("Usuário não encontrado")

    @strawberry.mutation
    def deletar_usuario(self, id: strawberry.ID) -> bool:
        global usuarios_db, playlists_db
        if not any(u.id == id for u in usuarios_db):
            raise Exception("Usuário não encontrado")
        usuarios_db = [u for u in usuarios_db if u.id != id]
        playlists_db[:] = [p for p in playlists_db if p.usuario_id != id]
        return True

    @strawberry.mutation
    def criar_musica(self, id: strawberry.ID, nome: str, artista: str) -> Musica:
        if any(m.id == id for m in musicas_db):
            raise Exception("Música já existe")
        m = Musica(id=id, nome=nome, artista=artista)
        musicas_db.append(m)
        return m

    @strawberry.mutation
    def atualizar_musica(self, id: strawberry.ID, nome: Optional[str] = None, artista: Optional[str] = None) -> Musica:
        for m in musicas_db:
            if m.id == id:
                if nome is not None:
                    m.nome = nome
                if artista is not None:
                    m.artista = artista
                return m
        raise Exception("Música não encontrada")

    @strawberry.mutation
    def deletar_musica(self, id: strawberry.ID) -> bool:
        global musicas_db, playlist_musica_db
        if not any(m.id == id for m in musicas_db):
            raise Exception("Música não encontrada")
        musicas_db = [m for m in musicas_db if m.id != id]
        playlist_musica_db[:] = [pm for pm in playlist_musica_db if pm[1] != id]
        return True

    @strawberry.mutation
    def criar_playlist(self, id: strawberry.ID, nome: str, usuario_id: strawberry.ID) -> Playlist:
        if not any(u.id == usuario_id for u in usuarios_db):
            raise Exception("Usuário não existe")
        if any(p.id == id for p in playlists_db):
            raise Exception("Playlist já existe")
        p = Playlist(id=id, nome=nome, usuario_id=usuario_id)
        playlists_db.append(p)
        return p

    @strawberry.mutation
    def atualizar_playlist(self, id: strawberry.ID, nome: Optional[str] = None) -> Playlist:
        for p in playlists_db:
            if p.id == id:
                if nome is not None:
                    p.nome = nome
                return p
        raise Exception("Playlist não encontrada")

    @strawberry.mutation
    def deletar_playlist(self, id: strawberry.ID) -> bool:
        global playlists_db, playlist_musica_db
        if not any(p.id == id for p in playlists_db):
            raise Exception("Playlist não encontrada")
        playlists_db = [p for p in playlists_db if p.id != id]
        playlist_musica_db[:] = [pm for pm in playlist_musica_db if pm[0] != id]
        return True

    @strawberry.mutation
    def adicionar_musica_playlist(self, playlist_id: strawberry.ID, musica_id: strawberry.ID) -> Playlist:
        if not any(p.id == playlist_id for p in playlists_db):
            raise Exception("Playlist não encontrada")
        if not any(m.id == musica_id for m in musicas_db):
            raise Exception("Música não encontrada")
        if any(pm == (playlist_id, musica_id) for pm in playlist_musica_db):
            raise Exception("Música já existe na playlist")
        playlist_musica_db.append((playlist_id, musica_id))
        return next(p for p in playlists_db if p.id == playlist_id)

    @strawberry.mutation
    def remover_musica_playlist(self, playlist_id: strawberry.ID, musica_id: strawberry.ID) -> Playlist:
        global playlist_musica_db
        if not any(p.id == playlist_id for p in playlists_db):
            raise Exception("Playlist não encontrada")
        if not any(m.id == musica_id for m in musicas_db):
            raise Exception("Música não encontrada")
        if not any(pm == (playlist_id, musica_id) for pm in playlist_musica_db):
            raise Exception("Música não está na playlist")
        playlist_musica_db = [pm for pm in playlist_musica_db if pm != (playlist_id, musica_id)]
        return next(p for p in playlists_db if p.id == playlist_id)

schema = strawberry.Schema(query=Query, mutation=Mutation)

from strawberry.fastapi import GraphQLRouter
graphql_app = GraphQLRouter(schema, path="/")
app = FastAPI(title="Streaming GraphQL API (Python)")
app.include_router(graphql_app)

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8002)